src/test/java/org/mvel2/compiler/MvelCompileExpNullSafeTest.java

```diff
@@ -120,6 +122,75 @@ public void testCustomResolverFactory() {
 
     }
 
+	/**
+	 * Tests NullSafe with a segment which shadows a variable.
+	 */
+	@Test
+	public void testNullSafeVariableOverlap() {
+		String expression = "foo.?bar.baz"; // "baz" shadows the variable named "baz".
+
+		final String expectedResult = "You got me!";
+
+		Map<String, Object> foo = new HashMap<>();
+		Map<String, Object> ctx = Collections.singletonMap("foo", foo);
+		Map<String, Object> bar = Collections.singletonMap("baz", expectedResult);
+		foo.put("bar", null); // start with null, as this will cause NullSafe to be added to the AST.
+
+		// This variable overlaps the segment to be resolved next after NullSafe.
+		Map<String, Object> vars = Collections.singletonMap("baz", "no, over here!");
+
+		Serializable compiledExpr = MVEL.compileExpression(expression);
+		Assert.assertNull(MVEL.executeExpression(compiledExpr, ctx, vars));
+
+		foo.put("bar", bar);
+		Assert.assertEquals(expectedResult, MVEL.executeExpression(compiledExpr, ctx, vars));
+	}
+
+	@Test
+	public void testNullSafeWithVariableAsMapKey() {
+		String expression = "foo.?bar[myvar]";
+
+		final String expectedResult = "You got me!";
+
+		Map<String, Object> foo = new HashMap<>();
+		Map<String, Object> ctx = Collections.singletonMap("foo", foo);
+		Map<String, Object> bar = Collections.singletonMap("baz", expectedResult);
+		foo.put("bar", null); // start with null, as this will cause NullSafe to be added to the AST.
+
+		Map<String, Object> vars = new HashMap<>();
+		vars.put("myvar", "baz");
+
+		Serializable compiledExpr = MVEL.compileExpression(expression);
+		Assert.assertNull(MVEL.executeExpression(compiledExpr, ctx, vars));
+
+		foo.put("bar", bar);
+		Assert.assertEquals(expectedResult, MVEL.executeExpression(compiledExpr, ctx, vars));
+	}
+
+	@Test
+	public void testNullSafeWithVariableAsMethodArg() {
+		String expression = "foo.?bar.convert(myvar)";
+
+		Map<String, Object> foo = new HashMap<>();
+		Map<String, Object> ctx = Collections.singletonMap("foo", foo);
+		foo.put("bar", null); // start with null, as this will cause NullSafe to be added to the AST.
+
+		Map<String, Object> vars = new HashMap<>();
+		vars.put("myvar", "hello");
+
+		Serializable compiledExpr = MVEL.compileExpression(expression);
+		Assert.assertNull(MVEL.executeExpression(compiledExpr, ctx, vars));
+
+		foo.put("bar", new StringConverter());
+		Assert.assertEquals("HELLO", MVEL.executeExpression(compiledExpr, ctx, vars));
+	}
```

Comment:
`testNullSafeWithVariableAsMapKey` and `testNullSafeWithVariableAsMethodArg` are additional tests to ensure that NullSafe can be used with variables.